cmd/kubeapps-apis/server/packages_test.go

```diff
@@ -0,0 +1,580 @@
+				mockedPackagingPlugin1,
+				mockedPackagingPlugin2,
+			},
+			request: &corev1.GetAvailablePackageSummariesRequest{
+				Context: &corev1.Context{
+					Cluster:   "",
+					Namespace: globalPackagingNamespace,
+				},
+				PaginationOptions: &corev1.PaginationOptions{PageToken: "3", PageSize: 1},
+			},
+
+			expectedResponse: &corev1.GetAvailablePackageSummariesResponse{
+				AvailablePackageSummaries: []*corev1.AvailablePackageSummary{
+					_test.MakeAvailablePackageSummary("pkg-2", mockedPackagingPlugin2.plugin),
+				},
+				Categories:    []string{"cat-1"},
+				NextPageToken: "4",
+			},
+			statusCode: codes.OK,
+		},
+		{
+			name: "it should successfully call and paginate (last page) the core GetAvailablePackageSummaries operation",
+			configuredPlugins: []*pkgsPluginWithServer{
+				mockedPackagingPlugin1,
+				mockedPackagingPlugin2,
+			},
+			request: &corev1.GetAvailablePackageSummariesRequest{
+				Context: &corev1.Context{
+					Cluster:   "",
+					Namespace: globalPackagingNamespace,
+				},
+				PaginationOptions: &corev1.PaginationOptions{PageToken: "3", PageSize: 1},
+			},
+
+			expectedResponse: &corev1.GetAvailablePackageSummariesResponse{
+				AvailablePackageSummaries: []*corev1.AvailablePackageSummary{
+					_test.MakeAvailablePackageSummary("pkg-2", mockedPackagingPlugin2.plugin),
+				},
+				Categories:    []string{"cat-1"},
+				NextPageToken: "4",
+			},
+			statusCode: codes.OK,
+		},
+		{
+			name: "it should successfully call and paginate (last page+1) the core GetAvailablePackageSummaries operation",
+			configuredPlugins: []*pkgsPluginWithServer{
+				mockedPackagingPlugin1,
+				mockedPackagingPlugin2,
+			},
+			request: &corev1.GetAvailablePackageSummariesRequest{
+				Context: &corev1.Context{
+					Cluster:   "",
+					Namespace: globalPackagingNamespace,
+				},
+				PaginationOptions: &corev1.PaginationOptions{PageToken: "4", PageSize: 1},
+			},
+
+			expectedResponse: &corev1.GetAvailablePackageSummariesResponse{
+				AvailablePackageSummaries: []*corev1.AvailablePackageSummary{},
+				Categories:                []string{"cat-1"},
+				NextPageToken:             "",
+			},
+			statusCode: codes.OK,
+		},
+		{
+			name: "it should fail when calling the core GetAvailablePackageSummaries operation with one of the plugin failing",
+			configuredPlugins: []*pkgsPluginWithServer{
+				mockedPackagingPlugin1,
+				mockedFailingPackagingPlugin,
+			},
+			request: &corev1.GetAvailablePackageSummariesRequest{
+				Context: &corev1.Context{
+					Cluster:   "",
+					Namespace: globalPackagingNamespace,
+				},
+			},
+
+			expectedResponse: &corev1.GetAvailablePackageSummariesResponse{
+				AvailablePackageSummaries: []*corev1.AvailablePackageSummary{},
+				Categories:                []string{""},
+			},
+			statusCode: codes.NotFound,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			server := &packagesServer{
+				plugins: tc.configuredPlugins,
+			}
```

Comment:
So this is fine as is, but just for a bit of explanation of what I'd meant in the previous review: If it's easy to configure the test plugin for each test, then each test above could instead be very explicit rather than referring to pre-built objects defined elsewhere in the file. I'll try to make some time to have a play later in the week perhaps and show you what I mean.

Reply:
It was fine as is, but thanks for updating. Note that you didn't need to add setters I don't think? What I was suggesting earlier is just making those fields (of the `TestPackagingPluginServer` as public fields (ie. capital first letter) so that call-sites can just set them. But fine with setters too. Still +1 :)

Reply:
> I consider using pre-built objects a more compact and readable was.
> But, on the other hand, it requires reading what the object/fn is really doing... tradeoffs

Yes, they are trade-offs. For tests, there is the point you mentioned (that you have to look elsewhere to see what the test input or output actually is), and also that you have to munge data into that one object for specific tests (like you were earlier for the `PkgVersion`), and that changing that one reference object effects many tests.

The opposite of having every object explicitly declared in each test also has its obvious downsides. I usually try for a good test object factory (simple make functions if i've nothing at hand), that enable defining just what is required to be read for the test with good defaults for the rest. But not always easy to get that balance.